### ISSUE TITLE
StackTraces are wrong in Tier-1 with tiered compilation enabled

### ISSUE THREAD

**Issue by Suchiman** [dotnet/runtime#88189]
### Description

Line numbers in StackTraces often start pointing to the closing curly brace of the method if the method gets uptiered.

### Reproduction Steps

```csharp
using System.Runtime.CompilerServices;

namespace BadStackTraces;

internal class Program
{
    static void Main(string[] args)
    {
        for (int i = 0; i < 100; i++)
        {
            Problem();
        }
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    public static void Problem()
    {
        try
        {
            GetEntities();
        }
        catch (Exception ex)
        {
            Console.WriteLine(ex.ToString());
        }
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    public static void GetEntities()
    {
        Thread.Sleep(1);
        throw new InvalidOperationException();
    }
}
```

### Expected behavior

The program always prints
```
System.InvalidOperationException: Operation is not valid due to the current state of the object.
   at BadStackTraces.Program.GetEntities() in \_\BadStackTraces\Program.cs:line 32
   at BadStackTraces.Program.Problem() in \_\BadStackTraces\Program.cs:line 20
```

### Actual behavior

After the `Problem` method recompiles for tier-1, the stacktrace points to the closing curly brace.
```
System.InvalidOperationException: Operation is not valid due to the current state of the object.
   at BadStackTraces.Program.GetEntities() in \_\BadStackTraces\Program.cs:line 32
   at BadStackTraces.Program.Problem() in \_\BadStackTraces\Program.cs:line 26
```

### Regression?

Most likely

### Known Workarounds

Adding `<TieredCompilation>false</TieredCompilation>` to the csproj makes the Problem go away.

### Configuration

.NET 7.0.8
Windows 10 / Azure App Service
x64

**Comment by jakobbotsch** [comment]
Doesn't repro with .NET 8. Looks like it was fixed by #74963 by this change:
![image](https://github.com/dotnet/runtime/assets/7887810/d63205e8-ad0d-47fd-9f9c-4729f0731d75)

**Comment by Suchiman** [comment]
> Closing as this is fixed in 8.0 via #74963

Any chance for a backport of these two lines?
